### PR TITLE
[OPTIMIZER] Added MeZO optimizer (backpropagation-free optimizer)

### PR DESCRIPTION
**Pull request by sachin-nntrainer**
MeZO is an Zeroth order optimizer based on the paper: [Fine-Tuning Language Models with Just Forward Passes](https://arxiv.org/abs/2305.17333)

Note: This PR has all the changes suggested in #3813 

This PR introduces MeZO (Memory-efficient Zeroth-Order) optimizer support to NNTrainer.  
It is the first backpropagation-free optimizer integrated into NNTrainer, enabling training without explicit gradient computation.

The implementation has been validated by training a fully connected (FC) network on the MNIST dataset within the NNTrainer framework.

### Motivation
Traditional optimizers in NNTrainer rely on backpropagation, which:
- Requires gradient storage
- Increases memory usage
- Can be inefficient for certain model classes or hardware constraints

MeZO provides an alternative by:
- Estimating gradients using zeroth-order optimization
- Eliminating the need for backpropagation
- Reducing memory footprint

This makes it particularly useful for:
- Memory-constrained environments
- On-device training scenarios
- Exploring gradient-free optimization techniques

### Accuracy Benchmarking
Dataset: MNIST (training sample: 60k, Validation sample: 10k)
Model: 3 fc layer with ReLU activation
Paramaters:  epochs=100 , MeZO_learning_rate=0.0001, MeZO_epsilon=0.001, batch_Size=32

Result:

Starting Training Loss: 0.687744
Starting Validation Loss: 0.506119
Accuracy: 18.0889%

Final training loss: 0.0897253
Final validation loss: 0.086979
Accuracy: 70.5729% 

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Sachin Singh <sachin.3@samsung.com>

### CODE REVIEW

**Comment by baek2sm** [comment]
**Review verdict: ✅ Approve** (rebased onto current `main`; all CI checks green)

### Summary
Adds the MeZO (Memory-efficient Zeroth-Order) backpropagation-free optimizer: a `requiresBackprop()` / `trainStep()` virtual interface on the optimizer base, the `mezo` implementation registered in `AppContext`, the `ML_TRAIN_OPTIMIZER_TYPE_MeZO` enum, a sample app, and unit tests. The branch was rebased onto the latest `upstream/main` (clean), preserving original authorship.

### Fixes applied on top to keep it main-compatible and CI-clean
- Guarded the `Applications/MeZO/jni` subdir against Tizen and Windows (the sample uses POSIX-only `arpa/inet.h`), matching the existing app-guard convention.
- Normalized the `trainStep()` empty-virtual brace style and the enum trailing-comment spacing for clang-format-14.
- Fixed a CodeQL integer-overflow finding by computing the sample-index offset in `size_t` (`Applications/MeZO/jni/main.cpp`).

### Stability for integration with latest `main`
The optimizer library uses only `<random>` and standard C++ (Windows-clean); the non-backprop train path is gated by `requiresBackprop()` so existing optimizers are unaffected. Windows MSVC build verified locally (504/504). All required checks are green.
